### ISSUE TITLE
Shuffle data in `iterative_train_test_split` beforehand (with solution)

### ISSUE THREAD

**Issue by erikhuck** [scikit-multilearn/scikit-multilearn#289]
Hello I've successfully forked the `scikit-multilearn` repo, branched off of master, and completed this issue but I was unable to make a pull request. Could I please be given write access or someone with write access can implement my solution.

The `iterative_train_test_split` function does not have a shuffle parameter and therefore does not shuffle the data ahead of time. I understand that the underlying `IterativeStratification` object does have a `shuffle` property which is set to `True` if a `random_state` is provided. But that doesn't shuffle because the underlying `_iter_test_indices` doesn't perform shuffling unlike the sibling class `KFold` which extends the same `_BaseKFold` parent class. The consequences of this are that when calling `iterative_train_test_split` across several cross-fold iterations, the instance-inclusion is not remotely normally distributed. Using an example dataset with ground truth including 12 labels (and about 6,000 instances), here's a plot of the instance-inclusion for the test set (20%`test_size`) distribution across 100 CV iterations:
<img width="599" alt="image" src="https://user-images.githubusercontent.com/42946548/229912675-38b75088-df29-476f-8e35-1504ff03159a.png">
What this plot means is that across the CV iterations, there were several of the test set instances that were included every iteration (all 100 of them). This means that the CV folds were not balanced enough because several of the instances were included every single time and the other instances hardly at all. **This means that when splitting the test dataset, we're mostly testing on the same instances over and over again rather than getting a balance of different instances in each fold**. The converse being that we're also training on the same instances over and over again rather than training on a balance of the instances from the complete dataset. After shuffling using my solution, the instance-inclusion distribution is much more normal:
<img width="596" alt="image" src="https://user-images.githubusercontent.com/42946548/229913439-535a20bc-dbbb-44fc-8d13-084a6b6768fb.png">
My solution additionally makes the deviation of label proportions in each fold from the original label proportions much smaller, improving the performance of the iterative stratification. Without shuffling, the test set proportions deviated by less than 5%
```
      "Label 1": 3.9850468575761893,
      "Label 2": 4.194394462998466,
      "Label 3": 4.462827268806928,
      "Label 4": 4.098111373194331,
      "Label 5": 3.7208254441759623,
      ...
```
But after shuffling the test set proportions deviated by less than 3%
```
      "Label 1": 2.1211386224760536,
      "Label 2": 2.0732622302730945,
      "Label 3": 2.0317471358023833,
      "Label 4": 2.8221312795976012,
      "Label 5": 2.370255420932089,
      ...
```
Here's my solution. Again I could just make a pull request since I already have the solution implemented but I need write access to this repo.
Change the function signature from this:
```
def iterative_train_test_split(X, y, test_size, random_state=None):
```
to this:
```
def iterative_train_test_split(X, y, test_size, random_state=None, shuffle=False):
```
Add a docstring description for the new parameter:
```
    shuffle : bool
        Whether to shuffle the data before splitting into batches. Note that the samples within each split
        will not be shuffled.
```
Add this code to the function before calling `next`:
```
    if shuffle:
        indices = list(y.index)
        check_random_state(random_state).shuffle(indices)
        X = X.loc[indices]
        y = y.loc[indices]
```

**Comment by erikhuck** [comment]
I haven't heard a response for a while. @ChristianSch I see you're a recent contributor. Are you able to either implement my solution or perhaps it'd be easier for you to give me write access so I can make a pull request?

**Comment by ChristianSch** [comment]
Hi! Thank you for making us aware. I didn't see the issue. You should be able to make a pull request without access. Do you get an error?

**Comment by erikhuck** [comment]
@ChristianSch Yes I do get this error when I try to create a new branch:
```
fatal: unable to access 'https://github.com/scikit-multilearn/scikit-multilearn.git/': The requested URL returned error: 403
```

I called the branch `shuffle` locally and the command I used was `git push origin shuffle`. I've never contributed to a repository that I wasn't an explicit member of with write access so I don't have experience with making a pull request as a non-member. Is there an additional step I'm missing? I know I can make a pull request. But I don't seem to be able to create a new remote branch in this repo. And I need to create the branch before I can make the pull request.

**Comment by ChristianSch** [comment]
Okay, got it. I think GitHub should present you with a big green button in your repo to create a PR. Otherwise you could try and see if GitHub knows about your branch anyhow on the pull requests page!

https://docs.github.com/articles/creating-a-pull-request-from-a-fork

**Comment by erikhuck** [comment]
@ChristianSch As seen in the page you linked, it says I need write access in order to make a pull request:
<img width="779" alt="image" src="https://user-images.githubusercontent.com/42946548/233387976-efed98fc-c91c-4b25-8553-774b0831aa78.png">
Are you able to give me write access?

**Comment by ChristianSch** [comment]
I could make you a contributor, but for the moment I would prefer not to to be honest. You need to create a pull request via the UI, not via the command line. We've had a multitude of pull requests from non-contributors so I'm a bit confused as to why you can't. Did you try via the UI?

You need to fork the project via the button, clone that forked repo, commit your changes there and then you should be able to create a pull request in the UI!

**Comment by erikhuck** [comment]
@ChristianSch Yes I have tried via the UI. My branch is not there. The name of my local branch is "shuffle" and as you can see, it's not an available branch
<img width="1204" alt="image" src="https://user-images.githubusercontent.com/42946548/233442503-b7cd8ada-fa17-48b0-b643-9fe960a20b3c.png">
It's because I can't push my changes. I'm not trying to create a pull request via the commandline. I already know that I can create a pull request via the UI. But I need a remote branch first. I'm only attempting to create my remote branch via the commandlne.

I used `git clone` on the commandline to fork the repo, I created a local branch via `git branch shuffle`, and then I made the necessary changes on the local branch and committed them via `git commit`. It's creating a remote branch that's the problem. I get the error message above when I do `git push origin shuffle`. I cannot create a pull request if my branch does not show up in the UI. You saw in the very documentation that you linked that it said write permissions are necessary to make a pull request. Unless you know of some other way.

**Comment by ChristianSch** [comment]
I think the problem is that you haven't properly forked the repo, as I don't see it in your repo list. You shouldn't clone this repo but your fork, that's what I meant ;) The button is on the top right.
